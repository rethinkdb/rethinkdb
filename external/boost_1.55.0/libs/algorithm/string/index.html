<html>
<head>
<meta http-equiv="refresh" content="0; URL=../../../doc/html/string_algo.html">
</head>
<body>
Automatic redirection failed, please go to
<a href="../../../doc/html/string_algo.html">../../doc/html/string_algo.html</a> 
&nbsp;<hr>
<p>© Copyright Beman Dawes, 2001</p>
<p>Distributed under the Boost Software License, Version 1.0. (See accompanying 
file <a href="../../../LICENSE_1_0.txt">LICENSE_1_0.txt</a> or copy 
at <a href="http://www.boost.org/LICENSE_1_0.txt">www.boost.org/LICENSE_1_0.txt</a>)</p>
</body>
</html>