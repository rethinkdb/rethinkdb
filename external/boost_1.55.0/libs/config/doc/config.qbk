[article Boost.Config
    [quickbook 1.4]
    [authors [Beman Dawes, Vesa Karvonen, John Maddock] ]
    [copyright 2001-2007 Beman Dawes, Vesa Karvonen, John Maddock]
    [category broken compiler workarounds]
    [id config]
    [dirname config]
    [purpose
        Helps boost library developers adapt to compiler idiosyncrasies; not intended for library users.
    ]
    [source-mode c++]
    [license
Distributed under the Boost Software License, Version 1.0.
(See accompanying file LICENSE_1_0.txt or copy at
[@http://www.boost.org/LICENSE_1_0.txt])
    ]
]


[/ Cited Boost resources ]

[def __BOOST_REGRESSION_TEST_DRIVER__ [@../../../../tools/regression/doc/index.html boost regression test driver]]
[def __BOOST_CONFIG_HEADER__ [@../../../../boost/config.hpp <boost/config.hpp>]]
[def __BOOST_CONFIG_USER_HEADER__ [@../../../../boost/config/user.hpp <boost/config/user.hpp>]]
[def __BOOST_CONFIG_SUFFIX_HEADER__ [@../../../../boost/config/user.hpp <boost/config/suffix.hpp>]]
[def __BOOST_CONFIG_DIR__ ['<boost-root>]`/boost/config/`]


[/ Other web resources ]

[def __STL_PORT__ [@http://stlport.sourceforge.net STLport]]
[def __BOOST_TRACKER__ [@https://svn.boost.org/trac/boost/newticket Trac]]
[def __CORE_LANGUAGE_DR337__ [@http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#337 Core Language DR337]]
[def __PRINCIPLES_AND_PATTERNS_ARTICLE__ [@http://www.objectmentor.com/resources/articles/Principles_and_Patterns.pdf following article]]


[/ Icons ]

[def __NOTE__ [$images/note.png]]
[def __ALERT__ [$images/caution.png]]
[def __DETAIL__ [$images/note.png]]
[def __TIP__ [$images/tip.png]]
[def __QUESTION_MARK__ [$images/question.png]]
[def __SPACE__ [$images/space.png]]
[def __GO_TO__ [$images/callouts/R.png]]


[/ Document files ]


[include configuring_boost.qbk]
[include macro_reference.qbk]
[include guidelines.qbk]
[include rationale.qbk]
[include acknowledgements.qbk]





