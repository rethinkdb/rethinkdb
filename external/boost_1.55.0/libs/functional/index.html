<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
<head>
  <meta http-equiv="Content-Language" content="en-us">
  <meta http-equiv="Content-Type" content="text/html; charset=us-ascii">
  <meta name="GENERATOR" content="Microsoft FrontPage 6.0">
  <meta name="ProgId" content="FrontPage.Editor.Document">

  <title>Boost Function Object Adapter Library</title>
</head>

<body bgcolor="#FFFFFF" text="#000000">
  <table border="1" bgcolor="#007F7F" cellpadding="2" summary="">
    <tr>
      <td bgcolor="#FFFFFF"><img src="../../boost.png" alt=
      "boost.png (6897 bytes)" width="277" height="86"></td>

      <td><a href="../../index.htm"><font face="Arial" color=
      "#FFFFFF"><big>Home</big></font></a></td>

      <td><a href="../libraries.htm"><font face="Arial" color=
      "#FFFFFF"><big>Libraries</big></font></a></td>

      <td><a href="http://www.boost.org/people/people.htm"><font face="Arial" color=
      "#FFFFFF"><big>People</big></font></a></td>

      <td><a href="http://www.boost.org/more/faq.htm"><font face="Arial" color=
      "#FFFFFF"><big>FAQ</big></font></a></td>

      <td><a href="../../more/index.htm"><font face="Arial" color=
      "#FFFFFF"><big>More</big></font></a></td>
    </tr>
  </table>

  <h1>Improved Function Object Adapters</h1>

  <p>The header <a href="../../boost/functional.hpp">functional.hpp</a>
  provides enhancements to the function object adapters specified in the C++
  Standard Library (sections 20.3.5, through to 20.3.8). The enhancements are
  principally possible due to two changes:</p>

  <ol>
    <li>We use the Boost <tt><a href=
    "../utility/call_traits.htm">call_traits</a></tt> templates to avoid the
    problem of <a href="binders.html#refref">references to references</a>,
    and to improve the efficiency of <a href="mem_fun.html#args">parameter
    passing</a>.</li>

    <li>We use two <a href="function_traits.html">function object traits</a>
    class templates to avoid the need for <tt><a href=
    "ptr_fun.html">ptr_fun</a></tt> with the adapters in this library.</li>
  </ol>

  <h3>Contents</h3>

  <p>The header contains the following function and class templates:</p>

  <table border="1" cellpadding="5" summary="">
    <tr>
      <th align="left"><a href="function_traits.html">Function object
      traits</a></th>

      <td valign="top"><tt>unary_traits<br>
      binary_traits</tt></td>

      <td valign="top">Used to determine the types of function objects' and
      functions' arguments. Eliminate the necessity for
      <tt>ptr_fun</tt>.</td>
    </tr>

    <tr>
      <th align="left"><a href="negators.html">Negators</a></th>

      <td valign="top"><tt>unary_negate<br>
      binary_negate<br>
      not1<br>
      not2</tt></td>

      <td valign="top">Based on section 20.3.5 of the standard.</td>
    </tr>

    <tr>
      <th align="left"><a href="binders.html">Binders</a></th>

      <td valign="top"><tt>binder1st<br>
      binder2nd<br>
      bind1st<br>
      bind2nd</tt></td>

      <td valign="top">Based on section 20.3.6 of the standard.</td>
    </tr>

    <tr>
      <th align="left"><a href="ptr_fun.html">Adapters for pointers to
      functions</a></th>

      <td valign="top"><tt>pointer_to_unary_function<br>
      pointer_to_binary_function<br>
      ptr_fun</tt></td>

      <td valign="top">Based on section 20.3.7 of the standard. Not required
      for use with this library since the binders and negators can adapt
      functions, but may be needed with third party adapters.</td>
    </tr>

    <tr>
      <th align="left"><a href="mem_fun.html">Adapters for pointers to member
      functions</a></th>

      <td valign="top"><tt>mem_fun_t<br>
      mem_fun1_t<br>
      const_mem_fun_t<br>
      const_mem_fun1_t<br>
      mem_fun_ref_t<br>
      mem_fun1_ref_t<br>
      const_mem_fun_ref_t<br>
      const_mem_fun1_ref_t<br>
      mem_fun<br>
      mem_fun_ref</tt></td>

      <td valign="top">Based on section 20.3.8 of the standard.</td>
    </tr>
  </table>

  <h3>Usage</h3>

  <p>Using these adapters should be pretty much the same as using the
  standard function object adapters; the only differences are that you need
  to write <tt>boost::</tt> instead of <tt>std::</tt>, and that you will get
  fewer headaches.</p>

  <p>For example, suppose you had a <tt>Person</tt> class that contained a
  <tt>set_name</tt> function:</p>

  <blockquote>
    <pre>
class Person
{
  public:
    void set_name(const std::string &amp;name);
  // ...
};
</pre>
  </blockquote>

  <p>You could rename a bunch of people in a collection, <tt>c</tt>, by
  writing</p>

  <blockquote>
    <pre>
std::for_each(c.begin(), c.end(), 
              boost::bind2nd(boost::mem_fun_ref(&amp;Person::set_name), "Fred"));
</pre>
  </blockquote>

  <p>If the standard adapters had been used instead then this code would
  normally fail to compile, because <tt>set_name</tt> takes a reference
  argument. Refer to the comments in the <a href="binders.html#refref">binder
  documentation</a> to explain why this is so.</p>

  <h3>Compiler Compatibility</h3>

  <p>The header and <a href="test/function_test.cpp">test program</a> have been
  compiled with the following compilers:</p>

  <table border="1" cellpadding="5" summary="">
    <tr>
      <th>Compiler</th>

      <th>Comments</th>
    </tr>

    <tr>
      <td valign="top">Borland C++Builder 4 Update 2</td>

      <td valign="top">No known issues.</td>
    </tr>

    <tr>
      <td valign="top">Borland C++ 5.5</td>

      <td valign="top">No known issues.</td>
    </tr>

    <tr>
      <td valign="top">g++ 2.95.2</td>

      <td valign="top">No known issues.</td>
    </tr>

    <tr>
      <td valign="top">Microsoft Visual C++ Service Pack 3</td>

      <td valign="top">
        Compiler lacks partial specialisation, so this library offers little
        more than is provided by the standard adapters:

        <ul>
          <li>The <tt>call_traits</tt> mechanism is unable to prevent
          references to references, and so the adapters in this library will
          be usable in fewer situations.</li>

          <li>The <tt>function_traits</tt> mechanism is unable to determine
          the argument and result types of functions, therefore
          <tt>ptr_fun</tt> continues to be required to adapt functions.</li>
        </ul>
      </td>
    </tr>
  </table>

  <h3>Future Directions</h3>

  <p>This library's primary focus is to solve the problem of references to
  references while maintaining as much compatibility as possible with the
  standard library. This allows you to use the techniques you read about in
  books and magazines with many of today's compilers.</p>

  <p>In the longer term, even better solutions are likely:</p>

  <ol>
    <li>Several Boost members are working on expression template libraries.
    These will allow a more natural syntax for combining and adapting
    functions. As this is a new technology, it may be some time before it has
    matured and is widely supported by major compilers but shows great
    promise. In the meantime, the functional.hpp library fills the gap.</li>

    <li>The Standard Committee has recognised the problem of references to
    references occurring during template instantiation and has moved to fix
    the standard (see the <a href=
    "http://anubis.dkuug.dk/jtc1/sc22/wg21/docs/cwg_active.html#106">C++
    standard core language active issues list</a>).</li>
  </ol>

  <h3>Author</h3>

  <p><a href="http://www.boost.org/people/mark_rodgers.htm">Mark Rodgers</a></p>

  <h3>Acknowledgements</h3>

  <p>Thanks to <a href="http://www.boost.org/people/john_maddock.htm">John Maddock</a> for
  suggesting the mechanism that allowed the function objects traits to work
  correctly. <a href="http://www.boost.org/people/jens_maurer.htm">Jens Maurer</a> provided
  invaluable feedback during the <a href=
  "http://www.boost.org/more/formal_review_process.htm">formal review process</a>.</p>
  <hr>

  <p><a href="http://validator.w3.org/check?uri=referer"><img border="0" src=
  "../../doc/images/valid-html401.png" alt="Valid HTML 4.01 Transitional"
  height="31" width="88"></a></p>

  <p>Revised 
  <!--webbot bot="Timestamp" s-type="EDITED" s-format="%d %B, %Y" startspan -->02
  December, 2006<!--webbot bot="Timestamp" endspan i-checksum="38510" --></p>

  <p><i>Copyright &copy; 2000 Cadenza New Zealand Ltd.</i></p>

  <p><i>Distributed under the Boost Software License, Version 1.0. (See
  accompanying file <a href="../../LICENSE_1_0.txt">LICENSE_1_0.txt</a> or
  copy at <a href=
  "http://www.boost.org/LICENSE_1_0.txt">http://www.boost.org/LICENSE_1_0.txt</a>)</i></p>
</body>
</html>
