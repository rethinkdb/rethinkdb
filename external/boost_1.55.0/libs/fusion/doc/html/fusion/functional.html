<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=US-ASCII">
<title>Functional</title>
<link rel="stylesheet" href="../../../../../doc/src/boostbook.css" type="text/css">
<meta name="generator" content="DocBook XSL Stylesheets V1.78.1">
<link rel="home" href="../index.html" title="Chapter&#160;1.&#160;Fusion 2.1">
<link rel="up" href="../index.html" title="Chapter&#160;1.&#160;Fusion 2.1">
<link rel="prev" href="extension/iterator_facade.html" title="Iterator Facade">
<link rel="next" href="functional/concepts.html" title="Concepts">
</head>
<body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF">
<table cellpadding="2" width="100%"><tr>
<td valign="top"><img alt="Boost C++ Libraries" width="277" height="86" src="../../../../../boost.png"></td>
<td align="center"><a href="../../../../../index.html">Home</a></td>
<td align="center"><a href="../../../../../libs/libraries.htm">Libraries</a></td>
<td align="center"><a href="http://www.boost.org/users/people.html">People</a></td>
<td align="center"><a href="http://www.boost.org/users/faq.html">FAQ</a></td>
<td align="center"><a href="../../../../../more/index.htm">More</a></td>
</tr></table>
<hr>
<div class="spirit-nav">
<a accesskey="p" href="extension/iterator_facade.html"><img src="../../../../../doc/src/images/prev.png" alt="Prev"></a><a accesskey="u" href="../index.html"><img src="../../../../../doc/src/images/up.png" alt="Up"></a><a accesskey="h" href="../index.html"><img src="../../../../../doc/src/images/home.png" alt="Home"></a><a accesskey="n" href="functional/concepts.html"><img src="../../../../../doc/src/images/next.png" alt="Next"></a>
</div>
<div class="section">
<div class="titlepage"><div><div><h2 class="title" style="clear: both">
<a name="fusion.functional"></a><a class="link" href="functional.html" title="Functional">Functional</a>
</h2></div></div></div>
<div class="toc"><dl class="toc">
<dt><span class="section"><a href="functional/concepts.html">Concepts</a></span></dt>
<dd><dl>
<dt><span class="section"><a href="functional/concepts/callable.html">Callable Object</a></span></dt>
<dt><span class="section"><a href="functional/concepts/reg_callable.html">Regular Callable
        Object</a></span></dt>
<dt><span class="section"><a href="functional/concepts/def_callable.html">Deferred Callable
        Object</a></span></dt>
<dt><span class="section"><a href="functional/concepts/poly.html">Polymorphic Function
        Object</a></span></dt>
</dl></dd>
<dt><span class="section"><a href="functional/invocation.html">Invocation</a></span></dt>
<dd><dl>
<dt><span class="section"><a href="functional/invocation/functions.html">Functions</a></span></dt>
<dt><span class="section"><a href="functional/invocation/metafunctions.html">Metafunctions</a></span></dt>
<dt><span class="section"><a href="functional/invocation/limits.html">Limits</a></span></dt>
</dl></dd>
<dt><span class="section"><a href="functional/adapters.html">Adapters</a></span></dt>
<dd><dl>
<dt><span class="section"><a href="functional/adapters/fused.html">fused</a></span></dt>
<dt><span class="section"><a href="functional/adapters/fused_procedure.html">fused_procedure</a></span></dt>
<dt><span class="section"><a href="functional/adapters/fused_function_object.html">fused_function_object</a></span></dt>
<dt><span class="section"><a href="functional/adapters/unfused.html">unfused</a></span></dt>
<dt><span class="section"><a href="functional/adapters/unfused_typed.html">unfused_typed</a></span></dt>
<dt><span class="section"><a href="functional/adapters/limits.html">Limits</a></span></dt>
</dl></dd>
<dt><span class="section"><a href="functional/generation.html">Generation</a></span></dt>
<dd><dl>
<dt><span class="section"><a href="functional/generation/functions.html">Functions</a></span></dt>
<dt><span class="section"><a href="functional/generation/metafunctions.html">Metafunctions</a></span></dt>
</dl></dd>
</dl></div>
<p>
      Components to call functions and function objects and to make Fusion code callable
      through a function object interface.
    </p>
<h4>
<a name="fusion.functional.h0"></a>
      <span class="phrase"><a name="fusion.functional.header"></a></span><a class="link" href="functional.html#fusion.functional.header">Header</a>
    </h4>
<pre class="programlisting"><span class="preprocessor">#include</span> <span class="special">&lt;</span><span class="identifier">boost</span><span class="special">/</span><span class="identifier">fusion</span><span class="special">/</span><span class="identifier">functional</span><span class="special">.</span><span class="identifier">hpp</span><span class="special">&gt;</span>
</pre>
<h4>
<a name="fusion.functional.h1"></a>
      <span class="phrase"><a name="fusion.functional.fused_and_unfused_forms"></a></span><a class="link" href="functional.html#fusion.functional.fused_and_unfused_forms">Fused
      and unfused forms</a>
    </h4>
<p>
      What is a function call?
    </p>
<pre class="programlisting"><span class="identifier">f</span> <span class="special">(</span><span class="identifier">a</span><span class="special">,</span><span class="identifier">b</span><span class="special">,</span><span class="identifier">c</span><span class="special">)</span>
</pre>
<p>
      It is a name and a tuple written next to each other, left-to-right.
    </p>
<p>
      Although the C++ syntax does not allow to replace <code class="literal">(a,b,c)</code>
      with some Fusion <a class="link" href="sequence.html" title="Sequence">Sequence</a>, introducing
      yet another function provides a solution:
    </p>
<pre class="programlisting"><span class="identifier">invoke</span><span class="special">(</span><span class="identifier">f</span><span class="special">,</span><span class="identifier">my_sequence</span><span class="special">)</span>
</pre>
<p>
      Alternatively it is possible to apply a simple transformation to <code class="literal">f</code>
      in order to achieve the same effect:
    </p>
<pre class="programlisting"><span class="identifier">f</span> <span class="identifier">tuple</span> <span class="special">&lt;=&gt;</span> f' <span class="special">(</span><span class="identifier">tuple</span><span class="special">)</span>
</pre>
<p>
      Now, <code class="literal">f'</code> is an unary function that takes the arguments to
      <code class="computeroutput"><span class="identifier">f</span></code> as a tuple; <code class="literal">f'</code>
      is the <span class="emphasis"><em>fused</em></span> form of <code class="computeroutput"><span class="identifier">f</span></code>.
      Reading the above equivalence right-to-left to get the inverse transformation,
      <code class="computeroutput"><span class="identifier">f</span></code> is the <span class="emphasis"><em>unfused</em></span>
      form of <code class="literal">f'</code>.
    </p>
<h4>
<a name="fusion.functional.h2"></a>
      <span class="phrase"><a name="fusion.functional.calling_functions_and_function_objects"></a></span><a class="link" href="functional.html#fusion.functional.calling_functions_and_function_objects">Calling
      functions and function objects</a>
    </h4>
<p>
      Having generic C++ code call back arbitrary functions provided by the client
      used to be a heavily repetitive task, as different functions can differ in
      arity, invocation syntax and other properties that might be part of the type.
      Transporting arguments as Fusion sequences and factoring out the invocation
      makes Fusion algorithms applicable to function arguments and also reduces the
      problem to one invocation syntax and a fixed arity (instead of an arbitrary
      number of arbitrary arguments times several syntactic variants times additional
      properties).
    </p>
<p>
      Transforming an unfused function into its fused counterpart allows n-ary calls
      from an algorithm that invokes an unary <a class="link" href="functional/concepts/poly.html" title="Polymorphic Function Object">Polymorphic
      Function Object</a> with <a class="link" href="sequence.html" title="Sequence">Sequence</a>
      arguments.
    </p>
<p>
      The library provides several function templates to invoke different kinds of
      functions and adapters to transform them into fused form, respectively. Every
      variant has a corresponding generator function template that returns an adapter
      instance for the given argument.
    </p>
<p>
      Constructors can be called applying <a href="http://www.boost.org/libs/functional/factory/doc/html/index.html" target="_top">Boost.Functional/Factory</a>.
    </p>
<h4>
<a name="fusion.functional.h3"></a>
      <span class="phrase"><a name="fusion.functional.making_fusion_code_callable_through_a_function_object_interface"></a></span><a class="link" href="functional.html#fusion.functional.making_fusion_code_callable_through_a_function_object_interface">Making
      Fusion code callable through a function object interface</a>
    </h4>
<p>
      Transforming a fused function into its unfused counterpart allows to create
      function objects to accept arbitrary calls. In other words, an unary function
      object can be implemented instead of (maybe heavily overloaded) function templates
      or function call operators.
    </p>
<p>
      The library provides both a strictly typed and a generic variant for this transformation.
      The latter should be used in combination with <a href="http://www.boost.org/libs/functional/forward/doc/html/index.html" target="_top">Boost.Functional/Forward</a>
      to attack <a href="http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2002/n1385.htm" target="_top">The
      Forwarding Problem</a>.
    </p>
<p>
      Both variants have a corresponding generator function template that returns
      an adapter instance for the given argument.
    </p>
</div>
<table xmlns:rev="http://www.cs.rpi.edu/~gregod/boost/tools/doc/revision" width="100%"><tr>
<td align="left"></td>
<td align="right"><div class="copyright-footer">Copyright &#169; 2001-2006, 2011, 2012 Joel de Guzman,
      Dan Marsden, Tobias Schwinger<p>
        Distributed under the Boost Software License, Version 1.0. (See accompanying
        file LICENSE_1_0.txt or copy at <a href="http://www.boost.org/LICENSE_1_0.txt" target="_top">http://www.boost.org/LICENSE_1_0.txt</a>)
      </p>
</div></td>
</tr></table>
<hr>
<div class="spirit-nav">
<a accesskey="p" href="extension/iterator_facade.html"><img src="../../../../../doc/src/images/prev.png" alt="Prev"></a><a accesskey="u" href="../index.html"><img src="../../../../../doc/src/images/up.png" alt="Up"></a><a accesskey="h" href="../index.html"><img src="../../../../../doc/src/images/home.png" alt="Home"></a><a accesskey="n" href="functional/concepts.html"><img src="../../../../../doc/src/images/next.png" alt="Next"></a>
</div>
</body>
</html>
