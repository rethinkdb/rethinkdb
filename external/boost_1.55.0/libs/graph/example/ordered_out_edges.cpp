//=======================================================================
// Copyright 1997, 1998, 1999, 2000 University of Notre Dame.
// Authors: Andrew Lumsdaine, Lie-Quan Lee, Jeremy G. Siek
//
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//=======================================================================
#include <boost/config.hpp>
#include <iostream>
#include <functional>
#include <string>

#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/properties.hpp>

/*
  Sample output:

  0  --chandler--> 1   --joe--> 1  
  1  --chandler--> 0   --joe--> 0   --curly--> 2   --dick--> 3   --dick--> 3  
  2  --curly--> 1   --tom--> 4  
  3  --dick--> 1   --dick--> 1   --harry--> 4  
  4  --tom--> 2   --harry--> 3  

  name(0,1) = chandler

  name(0,1) = chandler
  name(0,1) = joe

 */

template <class StoredEdge>
struct order_by_name
  : public std::binary_function<StoredEdge,StoredEdge,bool> 
{
  bool operator()(const StoredEdge& e1, const StoredEdge& e2) const {
    // Order by target vertex, then by name. 
    // std::pair's operator< does a nice job of implementing
    // lexicographical compare on tuples.
    return std::make_pair(e1.get_target(), boost::get(boost::edge_name, e1))
      < std::make_pair(e2.get_target(), boost::get(boost::edge_name, e2));
  }
};

#if !defined BOOST_NO_TEMPLATE_PARTIAL_SPECIALIZATION
struct ordered_set_by_nameS { };
namespace boost {
  template <class ValueType>
  struct container_gen<ordered_set_by_nameS, ValueType> {
    typedef std::multiset<ValueType, order_by_name<ValueType> > type;
  };
}
#else
struct ordered_set_by_nameS {
  template <class T>
  struct bind_ { typedef std::multiset<T, order_by_name<T> > type; };
};
namespace boost {
  template <> struct container_selector<ordered_set_by_nameS>  {
    typedef ordered_set_by_nameS type;
  };
}
#endif

namespace boost {
  template <>
  struct parallel_edge_traits<ordered_set_by_nameS> { 
    typedef allow_parallel_edge_tag type;
  };
}

int
main()
{
#ifdef BOOST_NO_TEMPLATE_PARTIAL_SPECIALIZATION
  std::cout << "This program requires partial specialization" << std::endl;
#else
  using namespace boost;
  typedef property<edge_name_t, std::string> EdgeProperty;
  typedef adjacency_list<ordered_set_by_nameS, vecS, undirectedS,
    no_property, EdgeProperty> graph_type;
  graph_type g;
  
  add_edge(0, 1, EdgeProperty("joe"), g);
  add_edge(1, 2, EdgeProperty("curly"), g);
  add_edge(1, 3, EdgeProperty("dick"), g);
  add_edge(1, 3, EdgeProperty("dick"), g);
  add_edge(2, 4, EdgeProperty("tom"), g);
  add_edge(3, 4, EdgeProperty("harry"), g);
  add_edge(0, 1, EdgeProperty("chandler"), g);

  property_map<graph_type, vertex_index_t>::type id = get(vertex_index, g);
  property_map<graph_type, edge_name_t>::type name = get(edge_name, g);

  graph_traits<graph_type>::vertex_iterator i, end;
  graph_traits<graph_type>::out_edge_iterator ei, edge_end;
  for (boost::tie(i, end) = vertices(g); i != end; ++i) {
    std::cout << id[*i] << " ";
    for (boost::tie(ei, edge_end) = out_edges(*i, g); ei != edge_end; ++ei)
      std::cout << " --" << name[*ei] << "--> " << id[target(*ei, g)] << "  ";
    std::cout << std::endl;
  }
  std::cout << std::endl;

  bool found;
  typedef graph_traits<graph_type> Traits;
  Traits::edge_descriptor e;
  Traits::out_edge_iterator e_first, e_last;

  boost::tie(e, found) = edge(0, 1, g);
  if (found)
    std::cout << "name(0,1) = " << name[e] << std::endl;
  else
    std::cout << "not found" << std::endl;
  std::cout << std::endl;

  boost::tie(e_first, e_last) = edge_range(0, 1, g);
  while (e_first != e_last)
    std::cout << "name(0,1) = " << name[*e_first++] << std::endl;
#endif
  return 0;
}
