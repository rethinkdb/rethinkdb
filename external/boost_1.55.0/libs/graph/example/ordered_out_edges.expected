0  --chandler--> 1   --joe--> 1  
1  --chandler--> 0   --joe--> 0   --curly--> 2   --dick--> 3   --dick--> 3  
2  --curly--> 1   --tom--> 4  
3  --dick--> 1   --dick--> 1   --harry--> 4  
4  --tom--> 2   --harry--> 3  

name(0,1) = chandler

name(0,1) = chandler
name(0,1) = joe
